[GStreamer][MSE] Extend text track support to MP4/3GPP timed text and CEA-608

vivienne-w opened the pull request:
#### 37fb0a0ead92c3e0470ee9bbf6fdedb9ea54f528
<pre>
[GStreamer][MSE] Extend text track support to MP4/3GPP timed text and CEA-608
<a href="https://bugs.webkit.org/show_bug.cgi?id=295634">https://bugs.webkit.org/show_bug.cgi?id=295634</a>

Reviewed by Philippe Normand and Alicia Boya Garcia.

Adds support for MPEG-4/3GPP timed text (known in MP4 as tx3g), as well as
experimental support for CEA-608 captions. Both formats are supported
by converting to WebVTT, using the optional parser/encoder elements in the
append pipeline.

tx3g support is straightforward - since qtdemux pushes raw text buffers,
we just need to plug webvttenc.

CEA-608 support rests on two elements: ccconverter and cea608tott.

As 608 captions are essentially a constant stream of byte pairs, translating
them into WebVTT is not quite so simple.
Specifically, cea608tott currently swallows the very last cue of the stream,
as without an EOS event it can never know at what point that caption is &quot;complete&quot;.

For experimental support, I consider this &quot;good enough for now&quot;.
Considerations for solving this in the future:
- append pipeline does not expect EOS events in normal operation, so while it would
  work to push the final cue, it will introduce much bigger issues
- instead, we could modify cea608tott itself to introduce a custom event, which would
  have the same effect on the element, without breaking the append pipeline
- the other issue is with timing of this event:
  - pushing it in response to endOfStream(), append pipeline would end up receiving a
    buffer outside of an append, which it can&apos;t handle
  - pushing it at each end-of-append will work, but will have the side effect
    of cutting any caption that stretches across multiple source buffers

Note also that currently, we expect the 608 captions to exist in the MP4 as a
separate track, just like other subtitle formats would, though it is more common
to see it inserted into the video stream. To extract those, we would need to additionally
plug the corresponding *ccextractor element after the video decoder.

That would require extra effort to handle this &quot;hidden&quot; track properly in AppendPipeline.

* LayoutTests/TestExpectations:
* LayoutTests/platform/glib/TestExpectations: Enable the new tests only for GTK + WPE ports.

* LayoutTests/media/media-source/content/counting-h264-mp4ttsimple-manifest.json: Added.
* LayoutTests/media/media-source/content/counting-h264-mp4ttsimple.mp4: Added.
* LayoutTests/media/media-source/media-source-mp4ttsimple-expected.txt: Added.
* LayoutTests/media/media-source/media-source-mp4ttsimple.html: Added.
  Same as media-source-vttsimple.

* LayoutTests/media/media-source/content/test-inband-cea608-track-manifest.json: Added.
* LayoutTests/media/media-source/content/test-inband-cea608-track.mp4: Added.
* LayoutTests/media/media-source/media-source-inband-cea608-track-expected.txt: Added.
* LayoutTests/media/media-source/media-source-inband-cea608-track.html: Added.
  Tests conversion to WebVTT. Note cea608tott only pushes 29 cues, instead of the expected 30 (see above).

* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
  Indicate support for tx3g and x-cea-608, if webvttenc and cea608tott respectively are available.
(WebCore::GStreamerRegistryScanner::ElementFactories::ElementFactories):
(WebCore::GStreamerRegistryScanner::ElementFactories::~ElementFactories):
(WebCore::GStreamerRegistryScanner::ElementFactories::elementFactoryTypeToString):
(WebCore::GStreamerRegistryScanner::ElementFactories::factory const):
(WebCore::GStreamerRegistryScanner::ElementFactories::hasElementForCaps const):
(WebCore::GStreamerRegistryScanner::initializeDecoders):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.h: Add ElementFactories::Type::CaptionEncoder

* Source/WebCore/platform/graphics/gstreamer/InbandTextTrackPrivateGStreamer.cpp: Fix some logging.
(WebCore::InbandTextTrackPrivateGStreamer::InbandTextTrackPrivateGStreamer):
(WebCore::InbandTextTrackPrivateGStreamer::notifyTrackOfSample):

* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::appsrcEndOfAppendCheckerProbe): Write pipeline dump at end-of-append.
(WebCore::AppendPipeline::parseDemuxerSrcPadCaps): Handle closedcaption/x-cea-608 as StreamType::Text
(WebCore::AppendPipeline::appsinkCapsChanged): Small fix for some &quot;incompatible caps&quot; errors I encountered.
(WebCore::createOptionalParserForFormat): Plug ccconverter for closedcaption/x-cea-608
(WebCore::createOptionalEncoderForFormat): Plug cea608tott for closedcaption/x-cea-608
(WebCore::AppendPipeline::Track::emplaceOptionalElementsForFormat): Combine both emplaceOptional... methods into one.
(WebCore::AppendPipeline::Track::emplaceOptionalParserForFormat): Deleted.
(WebCore::AppendPipeline::Track::emplaceOptionalEncoderForFormat): Deleted.
(WebCore::AppendPipeline::Track::initializeElements):
(WebCore::AppendPipeline::recycleTrackForPad):
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.h:

* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::enqueueSample): Log contents of text samples.
(WebCore::SourceBufferPrivateGStreamer::isReadyForMoreSamples): Fix some logging.

Canonical link: <a href="https://commits.webkit.org/298284@main">https://commits.webkit.org/298284@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91e90edeba0ed0026f3d7d386fb45cfd23d7ec50

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115008 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34724 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25205 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121128 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65658 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35369 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43285 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/87391 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117956 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28184 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103252 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67787 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27356 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21372 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64780 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97565 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21488 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124318 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41981 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/31384 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/96192 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42353 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99442 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95977 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24424 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41175 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19014 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41856 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47391 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41400 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44717 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43138 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/23e806ddbc6a2220e25f7ba6a7b74716af3b39c4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47763

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -1091,39 +1097,42 @@ void AppendPipeline::Track::emplaceOptionalParserForFormat(GstBin* bin, const GR
+    ASSERT(encoder);
+
+    gst_bin_add_many(bin, parser.get(), encoder.get(), nullptr);
+    gst_element_sync_state_with_parent(encoder.get());
     gst_element_sync_state_with_parent(parser.get());
```
These should be called after linking. I fixed it recently in other places of this file but forgot about this one.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e740dfa39f0982b4908899937fb843f5df0082ee. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47763

ntrrgc reviewed (approved):
LGTM.

vivienne-w committed:
[GStreamer][MSE] Extend text track support to MP4/3GPP timed text and CEA-608
https://bugs.webkit.org/show_bug.cgi?id=295634

Reviewed by Philippe Normand and Alicia Boya Garcia.

Adds support for MPEG-4/3GPP timed text (known in MP4 as tx3g), as well as
experimental support for CEA-608 captions. Both formats are supported
by converting to WebVTT, using the optional parser/encoder elements in the
append pipeline.

tx3g support is straightforward - since qtdemux pushes raw text buffers,
we just need to plug webvttenc.

CEA-608 support rests on two elements: ccconverter and cea608tott.

As 608 captions are essentially a constant stream of byte pairs, translating
them into WebVTT is not quite so simple.
Specifically, cea608tott currently swallows the very last cue of the stream,
as without an EOS event it can never know at what point that caption is "complete".

For experimental support, I consider this "good enough for now".
Considerations for solving this in the future:
- append pipeline does not expect EOS events in normal operation, so while it would
  work to push the final cue, it will introduce much bigger issues
- instead, we could modify cea608tott itself to introduce a custom event, which would
  have the same effect on the element, without breaking the append pipeline
- the other issue is with timing of this event:
  - pushing it in response to endOfStream(), append pipeline would end up receiving a
    buffer outside of an append, which it can't handle
  - pushing it at each end-of-append will work, but will have the side effect
    of cutting any caption that stretches across multiple source buffers

Note also that currently, we expect the 608 captions to exist in the MP4 as a
separate track, just like other subtitle formats would, though it is more common
to see it inserted into the video stream. To extract those, we would need to additionally
plug the corresponding *ccextractor element after the video decoder.

That would require extra effort to handle this "hidden" track properly in AppendPipeline.

* LayoutTests/TestExpectations:
* LayoutTests/platform/glib/TestExpectations: Enable the new tests only for GTK + WPE ports.

* LayoutTests/media/media-source/content/counting-h264-mp4ttsimple-manifest.json: Added.
* LayoutTests/media/media-source/content/counting-h264-mp4ttsimple.mp4: Added.
* LayoutTests/media/media-source/media-source-mp4ttsimple-expected.txt: Added.
* LayoutTests/media/media-source/media-source-mp4ttsimple.html: Added.
  Same as media-source-vttsimple.

* LayoutTests/media/media-source/content/test-inband-cea608-track-manifest.json: Added.
* LayoutTests/media/media-source/content/test-inband-cea608-track.mp4: Added.
* LayoutTests/media/media-source/media-source-inband-cea608-track-expected.txt: Added.
* LayoutTests/media/media-source/media-source-inband-cea608-track.html: Added.
  Tests conversion to WebVTT. Note cea608tott only pushes 29 cues, instead of the expected 30 (see above).

* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
  Indicate support for tx3g and x-cea-608, if webvttenc and cea608tott respectively are available.
(WebCore::GStreamerRegistryScanner::ElementFactories::ElementFactories):
(WebCore::GStreamerRegistryScanner::ElementFactories::~ElementFactories):
(WebCore::GStreamerRegistryScanner::ElementFactories::elementFactoryTypeToString):
(WebCore::GStreamerRegistryScanner::ElementFactories::factory const):
(WebCore::GStreamerRegistryScanner::ElementFactories::hasElementForCaps const):
(WebCore::GStreamerRegistryScanner::initializeDecoders):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.h: Add ElementFactories::Type::CaptionEncoder

* Source/WebCore/platform/graphics/gstreamer/InbandTextTrackPrivateGStreamer.cpp: Fix some logging.
(WebCore::InbandTextTrackPrivateGStreamer::InbandTextTrackPrivateGStreamer):
(WebCore::InbandTextTrackPrivateGStreamer::notifyTrackOfSample):

* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::appsrcEndOfAppendCheckerProbe): Write pipeline dump at end-of-append.
(WebCore::AppendPipeline::parseDemuxerSrcPadCaps): Handle closedcaption/x-cea-608 as StreamType::Text
(WebCore::AppendPipeline::appsinkCapsChanged): Small fix for some "incompatible caps" errors I encountered.
(WebCore::createOptionalParserForFormat): Plug ccconverter for closedcaption/x-cea-608
(WebCore::createOptionalEncoderForFormat): Plug cea608tott for closedcaption/x-cea-608
(WebCore::AppendPipeline::Track::emplaceOptionalElementsForFormat): Combine both emplaceOptional... methods into one.
(WebCore::AppendPipeline::Track::emplaceOptionalParserForFormat): Deleted.
(WebCore::AppendPipeline::Track::emplaceOptionalEncoderForFormat): Deleted.
(WebCore::AppendPipeline::Track::initializeElements):
(WebCore::AppendPipeline::recycleTrackForPad):
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.h:

* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::enqueueSample): Log contents of text samples.
(WebCore::SourceBufferPrivateGStreamer::isReadyForMoreSamples): Fix some logging.

Canonical link: https://commits.webkit.org/298284@main
